File: 백준/Gold/16227. 의약품 수송/의약품 수송.py
# https://www.acmicpc.net/problem/16227
import sys
from heapq import heappop, heappush
input = sys.stdin.readline
INF = float('inf')


def dijkstra():
    times = [[INF]*101 for _ in range(n+2)]
    times[0][0] = 0

    hq = []
    heappush(hq, (0, 0, 0))  # 총 시간, 안 쉬고 달려온 시간, 위치

    while hq:
        time, once, now = heappop(hq)
        if now == n+1:
            return time

        if times[now][once] < time:
            continue

        for nxt, nt in graph[now]:
            nxt_time = nt + time

            # 이번 노드에서 쉬고 다음 노드로 이동
            if times[nxt][nt] > nxt_time + 5:
                times[nxt][nt] = nxt_time + 5
                heappush(hq, (nxt_time+5, nt, nxt))

            # 쉬지 않고 바로 이동
            nxt_once = once + nt
            if nxt_once > 100:
                continue
            if times[nxt][nxt_once] > nxt_time:
                times[nxt][nxt_once] = nxt_time
                heappush(hq, (nxt_time, nxt_once, nxt))
    return INF


n, k = map(int, input().split())
graph = [[] for _ in range(n+2)]
for _ in range(k):
    u, v, t = map(int, input().split())
    if t > 100:
        continue
    graph[u].append((v, t))
    graph[v].append((u, t))

print(dijkstra())

File: 백준/Gold/16227. 의약품 수송/의약품 수송.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        List<int[]>[] graph = new ArrayList[n+2];
        for (int i = 0; i <= n+1; i++)
            graph[i] = new ArrayList<>();

        while (k-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int t = Integer.parseInt(st.nextToken());
            if (t > 100)
                continue;

            graph[u].add(new int[]{v, t});
            graph[v].add(new int[]{u, t});
        }

        System.out.println(dijkstra(graph, n));
    }

    private static int dijkstra(List<int[]>[] graph, int n) {
        int INF = Integer.MAX_VALUE;
        int[][] times = new int[n+2][101];
        for (int[] time: times)
            Arrays.fill(time, INF);

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparing(o -> o[0]));
        pq.add(new int[]{0, 0, 0});

        while (!pq.isEmpty()) {
            int[] first = pq.poll();
            int time = first[0];
            int once = first[1];
            int now = first[2];

            if (now == n+1)
                return time;
            if (times[now][once] < time)
                continue;

            for (int[] nxtNode: graph[now]) {
                int nxt = nxtNode[0];
                int nearOnce = nxtNode[1];
                int nxtTime = nearOnce + time;

                if (times[nxt][nearOnce] > nxtTime + 5) {
                    times[nxt][nearOnce] = nxtTime + 5;
                    pq.add(new int[]{nxtTime + 5, nearOnce, nxt});
                }

                int nxtOnce = once + nearOnce;
                if (nxtOnce <= 100 && times[nxt][nxtOnce] > nxtTime) {
                    times[nxt][nxtOnce] = nxtTime;
                    pq.add(new int[]{nxtTime, nxtOnce, nxt});
                }
            }
        }

        return -1;
    }
}